stablehlo/dialect/TypeInference.cpp

```diff
@@ -1114,20 +1103,26 @@ LogicalResult inferSliceOp(Optional<Location> location, Value operand,
   SmallVector<int64_t, 4> shape;
   shape.reserve(rank);
   SmallVector<int64_t> resultBounds;
+
+  auto inferSliceDim = [&](int64_t inputDim, int64_t start, int64_t end,
+                           int64_t stride) {
+    return inputDim == ShapedType::kDynamicSize
+               ? ShapedType::kDynamicSize
+               : static_cast<int64_t>(llvm::divideCeil(end - start, stride));
+  };
+
   for (int64_t i = 0, e = rank; i != e; i++) {
-    if (!inputBounds.empty())
-      resultBounds.push_back(
-          inferSliceDim(inputBounds[i], start[i], limit[i], strideVals[i]));
     if (hlo::isDynamicDimSize(rankedTy.getDimSize(i))) {
       shape.push_back(ShapedType::kDynamicSize);
-      continue;
+      if (inputBounds.empty()) continue;  // dynamic dim without bounds
```

Comment:
Could we move this at L1144? (have similar comment in the other PR.) I spoke too soon there that this PR logic seems good.

That way we check that start is valid even if we are not planning to use it in the computation.

Might be related to your plans to share verification between shape functions and op verify so it is okay if we just add TODO now.

Reply:
Please note that the "op verifier" thing has nothing to do with the topic we discussed here: the `PadOp::verify()` was totally removed a while ago.

"Could we move this at L1144",
"That way we check that start is valid even if we are not planning to use it in the computation." 
Let me explain this situation a little bit more: In existing StableHLO/MHLO code we have
```
for each dim:
  if (hlo::isDynamicDimSize(rankedTy.getDimSize(i))) {
      shape.push_back(ShapedType::kDynamicSize);
      continue;
  // Checks for attributes
  // calculate static result_dim
```
which actually tolerate any arbitrary attr values for a "?" dim: it does not check anything. 

Plan A (my current commit): keep tolerate invalid attr values and do "continue;" for unbounded case.
Plan B (Smit suggests, I like it but with concerns): Move the `// Checks for attributes` part firstly whatever.
@burmako What do you think of this (The other PR has exactly same issue) ? Like Smit says, we can "move this line at L1144", but in a (maybe small) risk that MHLO could have "positive tests with invalid attrs in '?' dim" ? Which may need more fixing in integration?

Reply:
ah, now I see that we rely on just shape inference for these ops. sorry for the confusion. These are already de-duped.

But, does that mean that the following op won't marked illegal?

"stablehlo.slice"(%arg0) {start_indices = dense<**[-1, 0]**> : tensor<2xi64>, limit_indices = dense<[2, 4]> : tensor<2xi64>, strides = dense<[1, 2]> : tensor<2xi64>} : (tensor<?x?xi32>) -> tensor<?x?xi32>
  %1 = "hlo_test_infer.get_return_type_components"(%0)


If that is the case, then Plan B becomes more important and it is better we do that sooner rather than later. I also doubt that we will hit many illegal IR as downstream users of these dialects will be checking that anyway. Just my hunch so it might still be better to verify upfront.

Reply:
I support Plan B - both here and for PadOp.

Reply:
As everyone is agreed (great!) Let's make the verification is correct order. And add a note in description that this maybe have MHLO impact.